[Linter] (Added linter into project)

AlwaysSayNo opened the pull request:
- Added golangci-lint linter into project;
- Created .golangci.yml file with custom setup of linter (default + additional lints);
- Refactored code (most and foremost err handling).

Commit:
[Linter] (Added linter into project)

- Added golangci-lint linter into project;
- Created .golangci.yml file with custom setup of linter (default + additional lints);
- Refactored code (most and foremost err handling).

Commit:
[Linter] (Added lint action)

- Added a GitHub action for linter.

Commit:
[Linter] (Added lint action - changes)

- Changed directory name.

Commit:
[Linter] (Latest golangci-lint for action)

- Installed latest golangci-lint for actions.

andriihar commented on .golangci.yml:
```diff
@@ -0,0 +1,20 @@
+linters:
+  enable:
+    # Default linters
+    - errcheck     # Detect unchecked errors
```
Here is a list of linters in our projects. Not all of them are very helpful, but you can add more. 

```
    - megacheck
    - govet
    - unconvert
    - megacheck
    - gas
    - gocyclo
    - dupl
    - misspell
    - unparam
    - typecheck
    - ineffassign
    - stylecheck
    - gochecknoinits
    - exportloopref
    - gocritic
    - revive
    - nakedret
    - gosimple
    - prealloc
    - unused
    - dogsled
    - gocognit
    - gofumpt
    - lll
    - noctx
   ```

AlwaysSayNo replied:
Replaced old linters with given, except two of them: 
- gofumpt. Reason: asserts "File is not `gofumpt`-ed" even for formatted files;
- noctx. Reason: cannot pass context into a scheduler properly.

AlwaysSayNo replied:
Is it OK?

andriihar replied:
gofumpt. Reason: asserts "File is not gofumpt-ed" even for formatted files;  - ok. 

* noctx. Reason: cannot pass context into a scheduler properly. - for this task, I think, it will be ok.

You can read about `graceful shutdown` in your app. In this case, you can pass the context to all workers.

andriihar commented on cmd/main.go:
```diff
@@ -5,16 +5,17 @@ import (
+	envs.Init()
 
-	dbUrl := db.GetUrl()
+	dbUrl := db.GetDatabaseUrl()
```
In Golang, we used UPPERCASE for all initialisms.

https://google.github.io/styleguide/go/decisions.html#initialisms

AlwaysSayNo replied:
With revive refactored all initialisms.

andriihar commented on internal/job/currency-job.go:
```diff
@@ -11,10 +12,19 @@ import (
+
+		err := currencyService.UpdateCurrencyRates()
+		if err != nil {
+			fmt.Printf("Error with: Update Currency Rates")
```
maybe here also a `log`?

Also, error string should not be capitalized or end with punctuation mark.

https://google.github.io/styleguide/go/decisions.html#error-strings

AlwaysSayNo replied:
Yes. Fixed it.

AlwaysSayNo replied:
Also used non-capputal letter for error messages, thx.

andriihar commented on internal/job/currency-job.go:
```diff
@@ -11,10 +12,19 @@ import (
+			log.Println("Finish job: Update Currency Rates")
+		}
 	})
+	if err != nil {
```
log error?

andriihar commented on internal/job/send-emails-job.go:
```diff
@@ -15,8 +15,14 @@ func SendEmailsJob(emailService *service.EmailService) {
-		log.Println("Start job: Send Emails")
+
+		err := emailService.SendEmails()
+		if err != nil {
```
```suggestion
		if err := emailService.SendEmails(); err != nil {
```

andriihar commented on internal/service/currency-service.go:
```diff
@@ -24,7 +24,11 @@ func NewCurrencyService() *CurrencyService {
-	c.UpdateCurrencyRates()
+
+	err := c.UpdateCurrencyRates()
+	if err != nil {
```
```suggestion
	if err := emailService.SendEmails(); err != nil {
```

andriihar commented on cmd/main.go:
```diff
@@ -29,5 +30,8 @@ func main() {
 	port := viper.Get("APP_PORT").(string)
-	r.Run(port)
+	err := r.Run(port)
+	if err != nil {
```
```suggestion
	if err := r.Run(port); err != nil {
```

andriihar commented on internal/service/currency-service.go:
```diff
@@ -100,14 +108,13 @@ func getApiUrl() string {
 	if err != nil {
-		log.Panic("Failed to update currency rates: ", err)
-		return
+		return err
```
You can wrap error here

https://preslav.me/2023/04/14/golang-error-handling-is-a-form-of-storytelling/
```suggestion
		return fmt.Errorf("getting currency rates: %w", err)
```

andiserg commented:
Good refactoring. There are no questions except those asked by @andriihar

Commit:
[Linter] (Used suggested linters)

- Used suggested from mentor linters.
- Fixed assertion from linter.

Commit:
[Linter] (Fixes while PR)

- Formatted error messages - now they start with non-capital letter;
- Replaced fmt with log for logging purposes.

andriihar commented:
As a future refactoring, I look towards graceful shutdown in applications. And then you will have the context of the execution of the entire apk, which you can send to any workers or jobs. and generally read about the topic of contexts in go.

Also,

```go
scheduler := cron.New(cron.WithLocation(time.UTC))
_, err := scheduler.AddFunc("0 * * * *", func() {
```

can be replaced by constructions of the type **for select with ticker**
```go
for {
   select {
    case <-ctx.Done():
        return
    case <-w.ticker.C:
       w.BestFuncEver() // some work
  }
}
```

Here some very simplified code from one of our projects.

```go
type SomeWatcher struct {
  fetchPeriodSec int

  ticker *time.Ticker
  close chan struct{}
}


func (w *SomeWatcher) Run(ctx context.Context) {
  w.ticker = time.NewTicker(time.Second * time.Duration(w.fetchPeriodSec))
  defer w.ticker.Stop()

  w.Sync()
  for {
  select {
  case <-ctx.Done():
  return
  case <-w.close:
  return
  case <-w.ticker.C:
  w.Sync() // some wotk
  }
  }
}
```

Then control over execution will also be at the level of the app and it will be possible to work with contexts to exit the worker when the app stops correctly.

And with `contexts` it will be possible to rework this part:

```go
resp, err := http.Get(apiURL.String())
```
